File: backend/src/patrones_de_diseño/multiple_dispatch/javascript/piedra_papel_tijeras.js
"use strict"

class Outcome {
	constructor(value, name){
		this.value = value
		this.name = name
	}

	toString(){
		return this.name;
	}

	static WIN(){ return new Outcome(0, "win") };
	static LOSE(){ return new Outcome(1, "lose") };
    static DRAW(){ return new Outcome(2, "win") };

}

Outcome.WIN = Outcome.WIN();
Outcome.LOSE = Outcome.LOSE();
Outcome.DRAW = Outcome.DRAW();


class Item {
	toString(){
		return this.constructor.name;
	}
};

class Paper extends Item {
	compete(item){ return item.evalPaper(); };
	evalPaper(item){ return Outcome.DRAW; };
	evalScissors(item){ return Outcome.WIN; };
	evalRock(item){ return Outcome.LOSE; };
}

class Scissors extends Item {
	compete(item) { return item.evalScissors(); };
	evalPaper(item) { return Outcome.LOSE; };
	evalScissors(item) { return Outcome.DRAW; };
	evalRock(item) { return Outcome.WIN };
}

class Rock extends Item {
	compete(item) { return item.evalRock(); };
	evalPaper(item) { return Outcome.WIN; };
	evalScissors(item) { return Outcome.LOSE; };
	evalRock(item) { return Outcome.DRAW; };
}



function match(item1, item2) { 
	var ready = item1 + " <--> " + item2 + " : ";
	console.log(ready + item1.compete(item2));
}

// Función para obtener un elemento al azar de un array
Array.prototype.sample = function() {
    return this[Math.floor(Math.random() * this.length)]
}

function itemPairGen(n) {
	var items = [Paper, Scissors, Rock], // Hardcoded :(
	    response = new Array;
	for (var i = 0; i < n; i++) {
		response.push([ items.sample(),
			            items.sample() ]
			          );
	};
	return response
};


/*
var matchs = itemPairGen(20);

console.log(matchs);

var item1 = matchs[0], item2 = matchs[1];

console.log(item1);
*/


for (var i = 0, matchs = itemPairGen(20); 
	     i < matchs.length; i++) {
    var item1 = matchs[i][0], item2 = matchs[i][1];
    //console.log(new item1);
    match(new item1, new item2);
};

File: backend/src/patrones_de_diseño/multiple_dispatch/ruby/piedra_papel_tijeras.rb
#!/usr/bin/ruby

class Outcome
	def initialize(value, name)
		@value = value
		@name = name
	end

	# Sobrecarga de operadores
	def to_s; return @name; end;

	def ==(other)
		return @value == other.value
	end

	# Métodos de clase
	def self.WIN; Outcome.new(0, "win"); end;
	def self.LOSE; Outcome.new(1, "lose"); end;
	def self.DRAW; Outcome.new(2, "draw"); end;
end


class Item
	def to_s; self.class.name; end;

	# Método para obtener las clases hijas
    def self.descendants
        ObjectSpace.each_object(Class).select { |klass| klass < self }
    end
end


class Paper < Item
	def compete(item)
		return item.evalPaper(self)
	end

	def evalPaper(item)
		return Outcome.DRAW
	end

	def evalScissors(item)
		return Outcome.WIN
	end

	def evalRock(item)
		return Outcome.LOSE
	end
end

class Scissors < Item
	def compete(item)
		return item.evalScissors(self)
	end

	def evalPaper(item)
		return Outcome.LOSE
	end

	def evalScissors(item)
		return Outcome.DRAW
	end

	def evalRock(item)
		return Outcome.WIN
	end
end

class Rock < Item
	def compete(item)
		return item.evalRock(self)
	end

	def evalPaper(item)
		return Outcome.WIN
	end

	def evalScissors(item)
		return Outcome.LOSE
	end

	def evalRock(item)
		return Outcome.DRAW
	end
end


def match(item1, item2)
	puts( "%s <--> %s: %s" % [
		item1, item2, item1.compete(item2)] )
end

def itemPairGen(n)
	response = []
	items = Item.descendants()
	for i in (1..n) do
        response.push( [items.sample.new,
                        items.sample.new] )
    end
    return response
end

for item1, item2 in itemPairGen(20) do
	match(item1, item2)
end
